src/geometric_algebra/nursery/chisolm.lean

```diff
@@ -69,12 +82,102 @@ local infix `*₊`:75 := sym_prod
 
 local infix `*₊ᵥ`:75 := sym_prod_vec
 
+
+def is_orthogonal (a : G₁) (b : G₁) : Prop := sym_prod_vec a b = 0
+
+theorem zero_is_orthogonal (a : G₁) : is_orthogonal 0 a := begin
+  unfold is_orthogonal,
+  unfold sym_prod_vec,
+  unfold prod_vec,
+  simp
+end
+
+/- a list of r orthogonal vectors, which may be non-unique -/
+def pairwise_ortho_vector (r : ℕ) := {l : vector G₁ r // l.val.pairwise is_orthogonal}
+
+/- need this for later, can't seem to get the type inference to work if I inline it-/
+-- def zero_pairwise_ortho_vector (r : ℕ) : pairwise_ortho_vector r := ⟨
+--   vector.repeat (0 : G₁) r, begin
+--   unfold vector.repeat subtype.val,
+--   apply pairwise_of_repeat,
+--   apply zero_is_orthogonal,
+-- end⟩
+
+
+-- r-blades
+def is_rblade (r : ℕ) (b : G) : Prop :=
+  -- a product of orthogonal vectors an a scalar
+  (∃ (k: G₀) (v : pairwise_ortho_vector r),
+   b = fₛ k * list.prod (v.val.val.map fᵥ))
+
+def Bᵣ (r : ℕ) := set_of (is_rblade r)
+
+namespace Bᵣ
+  variables {r : ℕ}
+
+  lemma all_G₀_is_rblade0 (k : G₀) : is_rblade 0 (fₛ k) := begin
+    use [k, list.pairwise.nil],
+    unfold vector.nil subtype.val list.map,
+    rw list.prod_nil,
+    simp,
+  end
+  lemma all_G₁_is_rblade1 (a : G₁) : is_rblade 1 (fᵥ a) := begin
+    use 1,
+    use ⟨(vector.cons a vector.nil), list.pairwise_singleton _ a⟩,
+    unfold vector.cons vector.nil subtype.val list.map,
+    rw [list.prod_cons, list.prod_nil],
+    simp,
+  end
+
+  instance has_coe_from_G₀ : has_coe G₀ (Bᵣ 0) := { coe := λ k, ⟨fₛ k, all_G₀_is_rblade0 k⟩}
+  instance has_coe_from_G₁ : has_coe G₁ (Bᵣ 1) := { coe := λ a, ⟨fᵥ a, all_G₁_is_rblade1 a⟩}
+
+  -- these are trivial, but somehow still needed
+  instance has_coe_to_G : has_coe (Bᵣ r) G := { coe := subtype.val }
+  @[simp]
+  lemma is_rblade (b : Bᵣ r) : is_rblade r b := b.property
+
+  /- todo: do we want this? -/
+  -- instance has_zero : has_zero (Bᵣ r) := {
+  --   zero := ⟨(0 : G), begin
+  --     use [0, zero_pairwise_ortho_vector r],
+  --     simp,
+  --   end⟩ 
+  -- }
+
+  def neg (b : Bᵣ r) : Bᵣ r := ⟨-b.val, begin
+    cases b with b' hb,
+    exact exists.elim hb begin
+      intros a ha,
+      use -a,
+      simp [ha],
+    end
+  end⟩
+
+  instance has_neg (r : ℕ) : has_neg (Bᵣ r) := { neg := neg}
+end Bᵣ
+
+-- r-vectors
+def Gᵣ (r : ℕ) := add_subgroup.closure (Bᵣ r)
+namespace Gᵣ
+  variables {r : ℕ}
+  instance addgroup : add_comm_group (Gᵣ r) := (Gᵣ r).to_add_comm_group
+end Gᵣ
+
+-- multi-vectors
+def Mᵣ (r : ℕ) := add_subgroup.closure (⋃ (r : ℕ), (Gᵣ r).carrier)
+namespace Mᵣ
+  variables {r : ℕ}
+  instance addgroup : add_comm_group (Mᵣ r) := (Mᵣ r).to_add_comm_group
+end Mᵣ
```

Comment:
I don't know how easy these will be to, say, show properties of multiplication on, but at least they were trivial to define

Reply:
I suspect the `to_add_comm_group` might already be inferred by lean.